src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilterBase.cs

```diff
@@ -98,10 +98,14 @@ private static void ValidateProperty(PropertyHelper propertyHelper)
 
             var propertyType = Nullable.GetUnderlyingType(property.PropertyType) ?? property.PropertyType;
 
-            if (!(propertyType.GetTypeInfo().IsPrimitive || propertyType == typeof(string)))
+            if (!TempDataSerializer.CanSerializeType(propertyType, out var actualType, out var errorMessage))
             {
-                throw new InvalidOperationException(
-                    Resources.FormatTempDataProperties_PrimitiveTypeOrString(property.DeclaringType.FullName, property.Name, nameof(TempDataAttribute)));
+                var messageWithPropertyInfo = Resources.FormatTempDataProperties_InvalidType(
+                    property.DeclaringType.FullName,
+                    property.Name,
+                    nameof(TempDataAttribute));
+
+                throw new InvalidOperationException($"{messageWithPropertyInfo} {errorMessage}");
```

Comment:
After this latest commit, do you still think we need a stringbuilder?

Reply:
nah this is OK now